README.md
# Resolve Open Alerts

A Datto RMM Component to resolve Open Alerts via the RMM API.

> This Component compliments existing Alert Resolution options in RMM, offering more flexibility and control over Alert Resolution, especially for 100s and 1000s of Open Alerts.

![License: GPL-3](https://img.shields.io/github/license/piouson/Resolve-Open-Alerts) ![Powershell: Version 4, 5 and 6](https://img.shields.io/badge/powershell-4.0%20%7C%205.1%20%7C%206.2-blue) ![Platforms: Windows, macOS, Linux](https://img.shields.io/badge/platform-windows%20%7C%20macos%20%7C%20linux-brightgreen)  
![Releave: Version 2](https://img.shields.io/github/v/release/piouson/Resolve-Open-Alerts?sort=semver) ![Release Date](https://img.shields.io/github/release-date/piouson/Resolve-Open-Alerts) ![Downloads](https://img.shields.io/github/downloads/piouson/Resolve-Open-Alerts/total) ![Open Issues](https://img.shields.io/github/issues-raw/piouson/Resolve-Open-Alerts)

> ![Resolve Open Alerts sample image](./sample-480.png)

## Prerequisites

- Valid RMM API token, saved on target device as `$Env:RMMAPIKey`, see [RMM API docs](https://help.aem.autotask.net/en/Content/2SETUP/APIv2.htm)
- Preconfigured [Environment Variables](#environment-variables)

## Deployment

### Import component in RMM

- Download [`resolve-open-alerts-v2.cpt`](https://github.com/piouson/Resolve-Open-Alerts/releases)
- [Import component in RMM](https://help.aem.autotask.net/en/Content/4WEBPORTAL/Components/ManageComponents.htm#Import_a_component), select `resolve-open-alerts-v2.cpt` and save

### Save RMM API token on target device

- In RMM, [Activate the API](https://help.aem.autotask.net/en/Content/2SETUP/APIv2.htm#Activate_the_API)
- Save the `API Key`, `API Secret Key` and `API URL` in a text file
- Download [RMM API example files](https://help.aem.autotask.net/en/Content/2SETUP/APIv2.htm#Downloadable_example_code) and extract `powershell_token.ps1`
- Open `Powershell ISE` as **`Administrator`** and open script `powershell_token.ps1`
- Scroll to the bottom of `powershell_token.ps1`, add code below and run script

> For options outside Powershell, use `curl_token.bat` or `csharp_token.cs`

```powershell
# Define parameters
$params = @{
  apiUrl        = '[API URL]'
  apiKey        = '[API Key]'
  apiSecretKey  = '[API Secret Key]'
}

# Call New-AemApiAccessToken function using defined parameters 
$key = New-AemApiAccessToken @params

# Save API Token as System Variable
[Environment]::SetEnvironmentVariable('RMMAPIKey',$key,'Machine')
```

> Requires process refresh for changes, close and reopen `Powershell ISE`, then run below command to confirm token saved as **System** variable

```powershell
[Environment]::GetEnvironmentVariable('RMMAPIKey', 'Machine')
```

> If token is not displayed, try saving again by rerunning `Powershell ISE` as **Administrator**
>
> **RMM API token** is not suitable as [RMM Site Variable](https://help.aem.autotask.net/en/Content/4WEBPORTAL/Sites/SiteSettings.htm#Variables) due to its character length exceeding limit.

### Run component on target device

- Use a [Quick Job](https://help.aem.autotask.net/en/Content/4WEBPORTAL/Jobs/Quick_Jobs.htm) or [Job Scheduler](https://help.aem.autotask.net/en/Content/4WEBPORTAL/Jobs/Job_Scheduler.htm) and select the target device

## Environment Variables

If running in production (e.g. from RMM), you only need to configure `$Env:RMMAPIKey` at **System** level, see [Deployment](#deployment).

Other variables required at deployment time, must be set manually in [`Invoke-MockComponent`](https://github.com/piouson/Resolve-Open-Alerts/blob/71b99a72c550e37e3bc72e8a6fd06ce743bd4083/ResolveAllOpenAlerts.ps1#L292) for local development and mock testing.

- `$Env:CS_WS_ADDRESS` - preconfigured in RMM {`merlot-centrastage.net` | `concord-centrastage.net` | etc}
- `$Env:Target` {`site` | `account`}, default is site of device running component
- `$Env:SiteID`, set here if `$Env:Target` = `"site"`
- `$Env:Priority` {`All` | `Information` | `Low` | `Moderate` | `High` | `Critical`}
- `$Env:MonitorType`, use Monitor Class names
- `$Env:DeviceType` {`Desktop` | `Laptop` | `Server` | `ESXI Host` | `Printer` | `Network Device`}
- `$Env:UdfNumber` {`1-30`}, UDF must be set to `"resolvealerts"` in RMM, see image below

> ![Sample UDF value](./udf-example.png)

See [variables in Powershell](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables) and [variables in RMM](https://help.aem.autotask.net/en/Content/2SETUP/AccountSettings/AccountSettings.htm#Variables).

## Running Directly in Powershell

The quickest way to try out this component is by running in Powershell.

- [Clone or Download source code](https://github.com/piouson/Resolve-Open-Alerts)
- Unzip to a suitable path
- Open `ResolveOpenAlerts.ps1` in [Visual Studio Code](https://code.visualstudio.com/), Powershell ISE or your favourite Powershell IDE.
- Define [environment variables](#environment-variables) in [`Invoke-MockComponent`](https://github.com/piouson/Resolve-Open-Alerts/blob/71b99a72c550e37e3bc72e8a6fd06ce743bd4083/ResolveAllOpenAlerts.ps1#L292) for testing or just above `Invoke-RMMComponent` for production/live
- For testing, switch to development mode, see [Running Tests](#running-tests)
- Run script

> The default RMM script timeout of 1 hour should resolve max 30,000 Alerts.

## Running Tests

> Requires [environment variables](#environment-variables).

Unit Tests have not been written for any functions yet, but [`Invoke-MockComponent`](https://github.com/piouson/Resolve-Open-Alerts/blob/71b99a72c550e37e3bc72e8a6fd06ce743bd4083/ResolveAllOpenAlerts.ps1#L292) can be used for local development, Integration Testing and/or Stress Testing. Just uncomment invocation at bottom of file, see snippet below.

> ### Production/Live mode

```powershell
#Invoke-MockComponent
Invoke-RMMComponent
```

> ### Development mode

```powershell
Invoke-MockComponent
#Invoke-RMMComponent
```

> When using [`Invoke-MockComponent`](https://github.com/piouson/Resolve-Open-Alerts/blob/71b99a72c550e37e3bc72e8a6fd06ce743bd4083/ResolveAllOpenAlerts.ps1#L292), Alerts are **never resolved**. Alert Resolution is simulated by using already Resolved Alerts as mock data.

For actual Alert Resolution in local development, use [`Invoke-RMMComponent`](https://github.com/piouson/Resolve-Open-Alerts/blob/71b99a72c550e37e3bc72e8a6fd06ce743bd4083/ResolveAllOpenAlerts.ps1#L261) directly with [environment variables](#environment-variables).

## Contributing

Anyone can contribute.

- Open a [new issue](https://github.com/piouson/Resolve-Open-Alerts/issues) to request bug fixes, features or a simple question
- Create a [pull request](https://github.com/piouson/Resolve-Open-Alerts/pulls) to submit your own fixes, unit tests, features, or for typos and document updates
- Submit test cases

## License

[![License: GPL-3](https://img.shields.io/github/license/piouson/Resolve-Open-Alerts)](https://github.com/piouson/Resolve-Open-Alerts/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## [2.3.4] - 23-05-2020

- Added Performance Counter Monitor to filter options
- Fixed issue with options filtering that causes random alert resolutions
- Fixed RMM component selection variable

## [2.0.0] - 21-05-2020

- Added development mode with mock testing
- Added more alert filtering options
- Changed code compatibility to Powershell v3, v5 and v6
- Changed alert filtering logic
- Changed API rate limit handling

## [1.2.1] - 27-11-2019

- Fixed alert resolution limit

## [1.2.0] - 02-11-2019

- Added API rate limit handling

## [1.1.0] - 01-11-2019

- Added API query control

## [1.0.0] - 31-10-2019

- Added initial release to GitHub


ResolveOpenAlerts.ps1
<#
    Resolve Open Alerts v2
    - Use Invoke-RMMComponent for production
    - Use Invoke-MockComponent for development, controls $script:isDevelopment
    - Development only simulates and Never resolves Alerts
    - You may use Invoke-RMMComponent in development to test actual Alert resoultion
    - isVerboseDetailed $true displays each Alert resolved in output
    - deviceCache is used to speedup device type filtering by skipping cached entries
    - For updates and documentation, see https://github.com/piouson/Resolve-Open-Alerts
    Author: piouson.github.io
#>

$script:version = " Resolve Open Alerts v2.3.9"
$script:apiHits = 0
$script:rateLimitCount = 0
$script:rateBuffer = 200
$script:delay = 30
$script:statusCode = 0
$script:isDevelopment = $false
$script:isVerboseDetailed = $true
$script:deviceCache = @{ }

function Test-ApiToken {
  return Test-Path Env:RMMAPIKey
}

function Test-RMMPlatform {
  return Test-Path Env:CS_WS_ADDRESS
}

function Get-RMMPlatform {
  [string]$platform = ($Env:CS_WS_ADDRESS -split '-')[0]
  return $platform
}

function Get-ApiUri {
  $platform = Get-RMMPlatform
  [string]$apiUri = "https://{0}-api.centrastage.net/api/v2/" -f $platform
  return $apiUri
}

function Get-ApiAlertUri {
  $alertPath = "/alerts/open"

  if ($script:isDevelopment) {
    $alertPath = "/alerts/resolved"
  }

  if ($Env:Target -eq "site") {
    $alertPath = "/{0}{1}" -f $Env:SiteID, $alertPath
  }
  [string]$alertApiUri = "{0}{1}{2}" -f (Get-ApiUri), $Env:Target, $alertPath
  return $alertApiUri
}

function Get-ApiDeviceUri {
  Param([string]$Uid)

  $devicePath = "device/{0}" -f $Uid
  [string]$deviceApiUri = "{0}{1}" -f (Get-ApiUri), $devicePath
  return $deviceApiUri
}

function Get-ApiHeader {
  $token = "Bearer {0}" -f $Env:RMMAPIKey
  return @{Authorization = $token }
}

function Show-ApiStatusError {
  Param([string]$Action)

  $message = switch ($script:statusCode) {
    401 { "[FAIL] Unauthorised Access! Check API token." }
    403 { "[FAIL] Access Denied! Check Security Level..." }
    404 { ("[FAIL] {0} Not Found!" -f $Action) }
    default { "[ERROR] Check Stderr for details..." }
  }
  return $message
}

function Test-RateLimit {
  Param(
    [string]$Hits,
    [string]$Buffer
  )

  return ($Hits -gt 0 -and $Hits % $Buffer -eq 0)
}

function Invoke-RMMApi {
  <#
.Description
RMM API limits requests to 600 per minute. Rate limiting begins at 500 requests.
This function invokes the RMM API ata rate of 200 requests every 30 seconds
#>
  Param(
    [Parameter(Mandatory = $true)]
    [string]$Uri,
    [Parameter(Mandatory = $false)]
    [string]$Method,
    [Parameter(Mandatory = $false)]
    [string]$Action
  )

  if (-not $Method) {
    $Method = "GET"
  }

  if (-not $Action) {
    $Action = "Alert"
  }

  if (Test-RateLimit -Hits $script:apiHits -Buffer $script:rateBuffer) {
    $script:rateLimitCount++
    Write-Verbose ("RateLimit x{0} | API hits: {1} | Sleep: {2}s" -f
      $script:rateLimitCount, $script:apiHits, $script:delay)
    Start-Sleep -Seconds $script:delay
  }

  try {
    $queryResults = Invoke-WebRequest -Uri $Uri -Headers (Get-ApiHeader) `
      -Method $Method -UseBasicParsing -Verbose:$false

    $script:apiHits++
    $script:statusCode = $_.Exception.Response.StatusCode.value__

  }
  catch {
    $script:statusCode = $_.Exception.Response.StatusCode.value__
    if ($script:isVerboseDetailed) {
      Write-Verbose (
        "{0} | {1}" -f (Show-ApiStatusError -Action $Action),
        ($Uri.split("api/")[1])
      )
    }
    if ($Action -eq "Alert") {
      Write-Error $_.Exception -ErrorAction Stop
    }
  }

  if ($queryResults) {
    return (ConvertFrom-Json $queryResults)
  }
  return $queryResults
}

function Get-OpenAlerts {
  Param([string]$Uri)

  return Invoke-RMMApi -Uri $Uri
}

function Get-Device {
  Param([string]$Uid)

  if ($script:deviceCache.Contains($Uid)) {
    return $script:deviceCache[$Uid]
  }
  $uri = Get-ApiDeviceUri -Uid $Uid
  $device = Invoke-RMMApi -Uri $uri -Action "Device"
  $script:deviceCache.Add($Uid, $device)
  return $device
}

function Find-AlertsByOptions {
  Param([PSCustomObject]$Alerts)

  if ($Env:Priority) {
    $Alerts.alerts = @(
      $Alerts.alerts | Where-Object {
        $_.priority -eq $Env:Priority
      }
    )
  }

  if ($Env:MonitorType -and $Alerts.alerts) {
    $Alerts.alerts = @(
      $Alerts.alerts | Where-Object {
        $_.alertContext."@class" -like "$Env:MonitorType*"
      }
    )
  }

  if ($Env:DeviceType -and $Alerts.alerts.alertSourceInfo) {
    $Alerts.alerts = @(
      $Alerts.alerts | Where-Object {(
        Get-Device $_.alertSourceInfo.deviceUid
      ).deviceType.category -like "$Env:DeviceType*"}
    )
  }

  if ($Env:UdfNumber -and $Alerts.alerts.alertSourceInfo) {
    $udf = ("udf{0}" -f $Env:UdfNumber)
    $Alerts.alerts = @(
      $Alerts.alerts | Where-Object {(
        Get-Device $_.alertSourceInfo.deviceUid
      ).udf.$udf -eq "resolvealerts"}
    )
  }

  $Alerts.pageDetails.count = $Alerts.alerts.count
  return $Alerts
}

function Resolve-OpenAlert {
  Param([string]$AlertUid)

  $resolvePath = "alert/{0}/resolve" -f $AlertUid
  $method = "POST"
  if ($script:isDevelopment) {

    $resolvePath = "alert/{0}" -f $AlertUid
    $method = "GET"
  }
  $alertUri = "{0}{1}" -f (Get-ApiUri), $resolvePath

  Invoke-RMMApi -Uri $alertUri -Method $method | Out-Null
}

function Get-FilterOptionsAsString {
  $message = ("[Filter Options]`n Target: {0}" -f
    (Get-Culture).TextInfo.ToTitleCase($Env:Target))
  if ($Env:Priority) {
    $message += ("`n Priority: {0}" -f $Env:Priority)
  }
  if ($Env:MonitorType) {
    $message += ("`n MonitorType: {0}" -f $Env:MonitorType)
  }
  if ($Env:DeviceType) {
    $message += ("`n DeviceType: {0}" -f $Env:DeviceType)
  }
  if ($Env:UdfNumber) {
    $message += ("`n UDF: {0}" -f $Env:UdfNumber)
  }
  return $message
}

function Resolve-AllAlerts {
  Param([string]$Uri)

  $openAlerts = @{ }
  $page = 0
  $resolvedCount = 0

  do {
    $nextPageUri = $openAlerts.pageDetails.nextPageUrl
    if ($nextPageUri) {
      $Uri = $nextPageUri
      $page++
      Write-Verbose ("NextPage: Page {0} | Total Alerts Resolved: {1}" -f
        $page, $resolvedCount)
    }

    $openAlerts = Get-OpenAlerts -Uri $Uri
    if (-not $openAlerts.alerts) {
      Write-Output "[Null] No Open Alerts for Account!"
      Write-Output ("[Stats] Total Alerts Resolved: {0}" -f $resolvedCount)
      return
    }

    $openAlerts = Find-AlertsByOptions -Alerts $openAlerts
    if (-not $openAlerts.alerts) {
      Write-Output "[Null] No Open Alerts matching filter options!"
    }
    else {
      Write-Output ("[New Page] Processing {0} Alert(s)..." -f $openAlerts.alerts.count)
      ForEach ($alert in $openAlerts.alerts) {
        Resolve-OpenAlert -AlertUid $alert.alertUid
        $resolvedCount++
        if ($script:isVerboseDetailed) {
          Write-Verbose ("Resolved Alert: {0} | Device {1}" -f
            $alert.alertUid, $alert.alertSourceInfo.deviceName)
        }
        if ($script:isDevelopment -and $resolvedCount -ge $script:maxSims) {
          Write-Output ("[Test Complete] Simulations: {0}" -f $resolvedCount)
          return
        }
      }
    }
    $nextPageUri = $openAlerts.pageDetails.nextPageUrl
  }
  While ($nextPageUri)

  Write-Output ("[Stats] Total Alerts Resolved: {0}" -f $resolvedCount)
}

function Invoke-RMMComponent {
  Write-Output "`n================================"
  Write-Output $script:version
  Write-Output "================================"

  Write-Output (Get-FilterOptionsAsString)

  if (-not (Test-ApiToken)) {
    Write-Output "[Auth] Token Error, view Stderr for details..."
    Write-Error "API Token Not Found!" -ErrorAction Stop
  }
  Write-Output "[Auth] Token found."

  if ((-not (Test-RMMPlatform)) -or ((Get-RMMPlatform).length -eq 0)) {
    Write-Output "[Platform] Not Found, view Stderr for details..."
    Write-Error "RMM Platform Unknown!" -ErrorAction Stop
  }
  Write-Output ("[Platform] RMM - {0}" -f (Get-RMMPlatform))

  Write-Output "[Fetch] Reading Open Alerts"
  Resolve-AllAlerts -Uri (Get-ApiAlertUri)
}

function Invoke-MockComponent {
  <#
.Description
Mock Run: Uses already Resolved Alerts for simulation
This is useful for Integration Testing and Stress Testing
Set Mock environment variables below
#>
  $Env:CS_WS_ADDRESS = "" # merlot-centrastage | concord-centrastage | etc
  $Env:RMMAPIKey = ""
  $Env:Target = "" # site | account
  $Env:SiteID = "" # set here if Env:Target = "site"
  $Env:Priority = "" # Information | Low | Moderate | High | Critical
  $Env:MonitorType = "" # online_offline | eventlog | custom_snmp | etc
  $Env:DeviceType = "" # Desktop | Laptop | Server | ESXI Host | Printer | etc
  $Env:UdfNumber = "" # UDF1-30, UDF must be set to "resolvealerts" in RMM
  $script:maxSims = 1000 # number of alert resolution to simulate

  $oldPrefs = $VerbosePreference
  $VerbosePreference = "Continue"
  $script:isDevelopment = $true

  Invoke-RMMComponent
  $VerbosePreference = $oldPrefs
}

#Invoke-MockComponent
Invoke-RMMComponent


ResolveAllOpenAlerts.ps1
<#
    Resolve Open Alerts v2.0.0
    - Use Invoke-RMMComponent for production
    - Use Invoke-MockComponent for development, controls $script:isDevelopment
    - Development only simulates and Never resolves Alerts
    - isVerboseDetailed $true shows each Alert resolved
    - DeviceCache to skip search for cached devices during Alert filtering
#>

$script:version = " Resolve All Open Alerts v1.4.2"
$script:apiHits = 0
$script:rateLimitCount = 0
$script:rateBuffer = 200
$script:delay = 30
$script:statusCode = 0
$script:isDevelopment = $false
$script:isVerboseDetailed = $true
$script:deviceCache = @{}

function Test-ApiToken {
	return Test-Path Env:RMMAPIKey
}

function Test-RMMPlatform {
    return Test-Path Env:CS_WS_ADDRESS
}

function Get-RMMPlatform {
    [string]$platform = ($Env:CS_WS_ADDRESS -split '-')[0]
    return $platform
}

function Get-ApiUri {
    $platform = Get-RMMPlatform
    [string]$apiUri = "https://{0}-api.centrastage.net/api/v2/" -f $platform
    return $apiUri
}

function Get-ApiAlertUri {
    $alertPath = "/alerts/open"

    if ($script:isDevelopment) {
        $alertPath = "/alerts/resolved"
    }

    if ($Env:Target -eq "site") {
        $alertPath = "/{0}{1}" -f $Env:SiteID, $alertPath
    }
    [string]$alertApiUri = "{0}{1}{2}" -f (Get-ApiUri), $Env:Target, $alertPath
    return $alertApiUri
}

function Get-ApiDeviceUri {
    Param([string]$Uid)

    $devicePath = "device/{0}" -f $Uid
    [string]$deviceApiUri = "{0}{1}" -f (Get-ApiUri), $devicePath
    return $deviceApiUri
}

function Get-ApiHeader {
    $token = "Bearer {0}" -f $Env:RMMAPIKey
    return @{Authorization = $token}
}

function Show-ApiStatusError {
    Param([string]$Action)

    $message = switch ($script:statusCode) {
        401 { "[FAIL] Unauthorised Access! Check API token." }
        403 { "[FAIL] Access Denied! Check Security Level..." }
        404 { ("[FAIL] {0} Not Found!" -f $Action) }
        default { "[ERROR] Check Stderr for details..." }
    }
    return $message
}

function Test-RateLimit {
    Param(
        [string]$Hits,
        [string]$Buffer
    )

    return ($Hits -gt 0 -and $Hits % $Buffer -eq 0)
}

function Invoke-RMMApi {
    <#
.Description
RMM API limits requests to 600 per minute. Rate limiting begins at 500 requests.
This function invokes the RMM API ata rate of 200 requests every 30 seconds
#>
    Param(
        [Parameter(Mandatory=$true)]
        [string]$Uri,
        [Parameter(Mandatory=$false)]
        [string]$Method,
        [Parameter(Mandatory=$false)]
        [string]$Action
    )

    if (-not $Method) {
        $Method = "GET"
    }

    if (-not $Action) {
        $Action = "Alert"
    }

    if (Test-RateLimit -Hits $script:apiHits -Buffer $script:rateBuffer) {
        $script:rateLimitCount++
        Write-Verbose ("RateLimit x{0} | API hits: {1} | Sleep: {2}s" -f
            $script:rateLimitCount, $script:apiHits, $script:delay)
        Start-Sleep -Seconds $script:delay
    }

    try {
        $queryResults = Invoke-WebRequest -Uri $Uri -Headers (Get-ApiHeader) `
            -Method $Method -UseBasicParsing -Verbose:$false

        $script:apiHits++
        $script:statusCode = $_.Exception.Response.StatusCode.value__

    } catch {
        $script:statusCode = $_.Exception.Response.StatusCode.value__
        if ($script:isVerboseDetailed) {
            Write-Verbose (
                "{0} | {1}" -f (Show-ApiStatusError -Action $Action),
                ($Uri.split("api/")[1])
            )
        }
        if ($Action -eq "Alert") {
            Write-Error $_.Exception -ErrorAction Stop
        }
    }

    if ($queryResults) {
        return (ConvertFrom-Json $queryResults)
    }
    return $queryResults
}

function Get-OpenAlerts {
    Param([string]$Uri)

    return Invoke-RMMApi -Uri $Uri
}

function Get-Device {
    Param([string]$Uid)

    if ($script:deviceCache.Contains($Uid)) {
        return $script:deviceCache.$Uid
    }
    $uri = Get-ApiDeviceUri -Uid $Uid
    $device = Invoke-RMMApi -Uri $uri -Action "Device"
    $script:deviceCache.Add($Uid, $device)
    return $device
}

function Find-AlertsByOptions {
    Param([PSCustomObject]$Alerts)

    if ($Env:Priority -ne 'All') {
        $Alerts.alerts = $Alerts.alerts | Where-Object {$_.priority -eq $Env:Priority}
    }

    if ($Env:MonitorType -and $Alerts.alerts.alertContext) {
        $Alerts.alerts = $Alerts.alerts | Where-Object {
            $_.alertContext."@class" -like "$Env:MonitorType*"
        }
    }

    if ($Env:DeviceType -and $Alerts.alerts.alertSourceInfo) {
        $Alerts.alerts = $Alerts.alerts | Where-Object {
            $deviceUid = $_.alertSourceInfo.deviceUid
            $device = Get-Device -Uid $deviceUid
            $device.deviceType.category -like "$Env:DeviceType*"
        }
    }

    if ($Env:UdfNumber -and $Alerts.alerts.alertSourceInfo) {
        $Alerts.alerts = $Alerts.alerts | Where-Object {
            $deviceUid = $_.alertSourceInfo.deviceUid
            $device = Get-Device -Uid $deviceUid
            $device.udf[$Env:UdfNumber] -eq "resolvealerts"
        }
    }

    return $Alerts
}

function Resolve-OpenAlert {
    Param([string]$AlertUid)

    $resolvePath = "alert/{0}/resolve" -f $AlertUid
    $method = "POST"
    if ($script:isDevelopment) {

        $resolvePath = "alert/{0}" -f $AlertUid
        $method = "GET"
    }
    $alertUri = "{0}{1}" -f (Get-ApiUri), $resolvePath

    Invoke-RMMApi -Uri $alertUri -Method $method | Out-Null
}

function Resolve-AllAlerts {
    Param([string]$Uri)

    $openAlerts = @{}
    $page = 0
    $resolvedCount = 0

    do {
        $nextPageUri = $openAlerts.pageDetails.nextPageUrl
        if ($nextPageUri) {
            $Uri = $nextPageUri
            $page++
            Write-Verbose ("NextPage: Page {0} | Total Alerts Resolved: {1}" -f
                $page, $resolvedCount)
        }

        $openAlerts = Get-OpenAlerts -Uri $Uri
        if (-not $openAlerts.alerts) {
            Write-Host "[FAIL] Error reading Alerts, see Stderr..."
            Write-Host (" Total Alerts Resolved: {0}" -f $resolvedCount)
            Write-Error "Open Alerts Not Found!" -ErrorAction Stop
        }

        $openAlerts = Find-AlertsByOptions -Alerts $openAlerts
        $alertCount = $openAlerts.alerts.count

        if ($alertCount -gt 0) {
            Write-Host ("[Next Page] Processing {0} Alert(s)..." -f $alertCount)
            ForEach ($alert in $openAlerts.alerts) {
                Resolve-OpenAlert -AlertUid $alert.alertUid
                $resolvedCount++
                if ($script:isVerboseDetailed) {
                    Write-Verbose ("Resolved Alert: {0} | Device {1}" -f
                        $alert.alertUid, $alert.alertSourceInfo.deviceName)
                }
                if ($script:isDevelopment -and $resolvedCount -ge $script:maxSims) {
                    Write-Host ("[Test Complete] Simulations: {0}" -f $resolvedCount)
                    return
                }
            }
            $nextPageUri = $openAlerts.pageDetails.nextPageUrl
        }
        else {
            if ($script:isVerboseDetailed) {
                Write-Warning "[No Data] No matching open Alerts..."
            }
        }
    }
    While($nextPageUri)

    Write-Verbose (" Total Alerts Resolved: {0}" -f $resolvedCount)
}

function Invoke-RMMComponent {
    Write-Host "`n================================"
    Write-Host $script:version
    Write-Host "================================"

    Write-Host "[Options]"
    Write-Host (" Target: {0}" -f (Get-Culture).TextInfo.ToTitleCase($Env:Target))
    if ($Env:Target -eq "site") {
        Write-Host (" SiteID: {0}" -f $Env:SiteID)
    }
    Write-Host (" Priority: {0}" -f $Env:Priority)
    Write-Host (" MonitorType: {0}" -f $Env:MonitorType)
    Write-Host (" DeviceType: {0}" -f $Env:DeviceType)
    Write-Host (" UDF: {0}" -f $Env:UdfNumber)

    if(-not (Test-ApiToken)) {
        Write-Host "[Auth] Token Error, view Stderr for details..."
        Write-Error "API Token Not Found!" -ErrorAction Stop
    }
    Write-Host "[Auth] Token found."

    if((-not (Test-RMMPlatform)) -or ((Get-RMMPlatform).length -eq 0)) {
        Write-Host "[Platform] Not Found, view Stderr for details..."
        Write-Error "RMM Platform Unknown!" -ErrorAction Stop
    }
    Write-Host ("[Platform] RMM - {0}" -f (Get-RMMPlatform))

    Write-Host "[Fetch] Reading Open Alerts"
    Resolve-AllAlerts -Uri (Get-ApiAlertUri)
}

function Invoke-MockComponent {
    <#
.Description
Mock Run: Only uses already Resolved Alerts for simulation
This is useful for Integration Testing and Stress Testing
Set Mock environment variables below
#>
    $Env:CS_WS_ADDRESS = "" # merlot-centrastage | concord-centrastage | etc
    $Env:RMMAPIKey = ""
    $Env:Target = "account" # site | account
    $Env:SiteID = "" # set here if Env:Target = "site"
    $Env:Priority = "All" # All | Information | Low | Moderate | High | Critical
    $Env:MonitorType = "" # online_offline | eventlog | custom_snmp | etc
    $Env:DeviceType = "" # Desktop | Laptop | Server | ESXI Host | Printer | etc
    $Env:UdfNumber = "" # UDF1-30, UDF must be set to "resolvealerts" in RMM
    $script:maxSims = 1000 # number of alert resolution to simulate

    $oldPrefs = $VerbosePreference
    $VerbosePreference = "Continue"
    $script:isDevelopment = $true

    Invoke-RMMComponent
    $VerbosePreference = $oldPrefs
}

#Invoke-MockComponent
Invoke-RMMComponent
